internal/client/db/db.go

```diff
@@ -0,0 +1,585 @@
+	default:
+		panic("unsupported block id type")
+	}
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) BlockNumberFromID(id generic.BlockID) (*N, error) {
+	switch id := id.(type) {
+	case generic.BlockIDHash[H]:
+		return bdb.Number(id.Inner)
+	case generic.BlockIDNumber[N]:
+		return &id.Inner, nil
+	default:
+		panic("unsupported block id type")
+	}
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) Body(hash H) ([]runtime.Extrinsic, error) {
+	bdb.pinnedBlocksCacheMtx.RLock()
+	defer bdb.pinnedBlocksCacheMtx.RUnlock()
+	body := bdb.pinnedBlocksCache.Body(hash)
+	if body != nil {
+		return *body, nil
+	}
+
+	return bdb.bodyUncached(hash)
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) Justifications(hash H) (runtime.Justifications, error) {
+	bdb.pinnedBlocksCacheMtx.RLock()
+	defer bdb.pinnedBlocksCacheMtx.RUnlock()
+	justifications := bdb.pinnedBlocksCache.Justifications(hash)
+	if justifications != nil {
+		return justifications, nil
+	}
+
+	return bdb.justificationsUncached(hash)
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) LastFinalized() (H, error) {
+	bdb.metaMtx.RLock()
+	defer bdb.metaMtx.RUnlock()
+	return bdb.meta.FinalizedHash, nil
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) Leaves() ([]H, error) {
+	bdb.leavesMtx.RLock()
+	defer bdb.leavesMtx.RUnlock()
+	return bdb.leaves.Hashes(), nil
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) DisplacedLeavesAfterFinalizing(blockNumber N) ([]H, error) {
+	bdb.leavesMtx.RLock()
+	defer bdb.leavesMtx.RUnlock()
+	return bdb.leaves.DisplacedByFinalHeight(blockNumber).Leaves(), nil
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) Children(parentHash H) ([]H, error) {
+	return readChildren[H](bdb.db, columns.Meta, metakeys.ChildrenPrefix, parentHash)
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) LongestContaining(baseHash H, importLock *sync.RWMutex) (*H, error) {
+	baseHeader, err := bdb.Header(baseHash)
+	if err != nil {
+		return nil, err
+	}
+	if baseHeader == nil {
+		return nil, nil
+	}
+
+	var getLeaves = func() ([]H, error) {
+		// ensure no blocks are imported during this code block.
+		// an import could trigger a reorg which could change the canonical chain.
+		// we depend on the canonical chain staying the same during this code block.
+		importLock.RLock()
+		defer importLock.RUnlock()
+		info := bdb.Info()
+		if info.FinalizedNumber > baseHeader.Number() {
+			// `baseHeader` is on a dead fork.
+			return nil, nil
+		}
+		return bdb.Leaves()
+	}
+
+	leaves, err := getLeaves()
+	if err != nil {
+		return nil, err
+	}
+
+	// for each chain. longest chain first. shortest last
+	for _, leafHash := range leaves {
+		currentHash := leafHash
+		// go backwards through the chain (via parent links)
+		for {
+			if currentHash == baseHash {
+				return &leafHash, nil
+			}
+
+			currentHeader, err := bdb.Header(currentHash)
+			if err != nil {
+				return nil, err
+			}
+			if currentHeader == nil {
+				return nil, fmt.Errorf("Failed to get header for hash %v", currentHash)
```

Comment:
General comment: Need to update error messages to be compatible with the linter

Reply:
🤔 I don't get any lint issues? The lint check workflow is green.

Reply:
oh sorry its just my IDE complaining, my bad